File: frontend/src/utils/logger.ts
/**
 * Comprehensive Logging System for Store Frontend
 * ================================================
 * 
 * Features:
 * - Multiple log levels with color coding
 * - API error logging
 * - User action tracking
 * - Performance monitoring
 * - Error boundary integration
 * - Console-only output (no file storage)
 */

export enum LogLevel {
  DEBUG = 'DEBUG',
  INFO = 'INFO',
  WARN = 'WARN',
  ERROR = 'ERROR',
  CRITICAL = 'CRITICAL'
}

interface LogEntry {
  timestamp: string;
  level: LogLevel;
  message: string;
  context?: Record<string, any>;
  error?: {
    name: string;
    message: string;
    stack?: string;
  };
  user?: {
    id?: number;
    username?: string;
  };
  page?: {
    url: string;
    title: string;
  };
}

class Logger {
  private name: string;
  private minLevel: LogLevel;
  
  // ANSI color codes for console
  private static COLORS = {
    DEBUG: '\x1b[36m',      // Cyan
    INFO: '\x1b[32m',       // Green
    WARN: '\x1b[33m',       // Yellow
    ERROR: '\x1b[31m',      // Red
    CRITICAL: '\x1b[35m',   // Magenta
    RESET: '\x1b[0m'        // Reset
  };
  
  private static LEVEL_PRIORITY = {
    DEBUG: 0,
    INFO: 1,
    WARN: 2,
    ERROR: 3,
    CRITICAL: 4
  };
  
  constructor(name: string = 'store-frontend', minLevel: LogLevel = LogLevel.INFO) {
    this.name = name;
    this.minLevel = minLevel;
  }
  
  /**
   * Build structured log entry
   */
  private buildLogEntry(
    level: LogLevel,
    message: string,
    context?: Record<string, any>,
    error?: Error
  ): LogEntry {
    const entry: LogEntry = {
      timestamp: new Date().toISOString(),
      level,
      message,
      context,
      page: {
        url: window.location.href,
        title: document.title
      }
    };
    
    // Add user info if available (from localStorage/sessionStorage)
    const userStr = localStorage.getItem('user');
    if (userStr) {
      try {
        const user = JSON.parse(userStr);
        entry.user = {
          id: user.id,
          username: user.username
        };
      } catch {}
    }
    
    // Add error info
    if (error) {
      entry.error = {
        name: error.name,
        message: error.message,
        stack: error.stack
      };
    }
    
    return entry;
  }
  
  /**
   * Check if log level should be output
   */
  private shouldLog(level: LogLevel): boolean {
    return Logger.LEVEL_PRIORITY[level] >= Logger.LEVEL_PRIORITY[this.minLevel];
  }
  
  /**
   * Output log to console with formatting
   */
  private output(entry: LogEntry) {
    if (!this.shouldLog(entry.level)) {
      return;
    }
    
    const color = Logger.COLORS[entry.level];
    const reset = Logger.COLORS.RESET;
    
    const prefix = `${color}[${entry.level}]${reset} ${entry.timestamp} | ${this.name} |`;
    
    // Use appropriate console method
    switch (entry.level) {
      case LogLevel.DEBUG:
        console.debug(prefix, entry.message, entry);
        break;
      case LogLevel.INFO:
        console.info(prefix, entry.message, entry);
        break;
      case LogLevel.WARN:
        console.warn(prefix, entry.message, entry);
        break;
      case LogLevel.ERROR:
      case LogLevel.CRITICAL:
        console.error(prefix, entry.message, entry);
        break;
    }
  }
  
  /**
   * Log debug message
   */
  debug(message: string, context?: Record<string, any>) {
    const entry = this.buildLogEntry(LogLevel.DEBUG, message, context);
    this.output(entry);
  }
  
  /**
   * Log info message
   */
  info(message: string, context?: Record<string, any>) {
    const entry = this.buildLogEntry(LogLevel.INFO, message, context);
    this.output(entry);
  }
  
  /**
   * Log warning message
   */
  warn(message: string, context?: Record<string, any>) {
    const entry = this.buildLogEntry(LogLevel.WARN, message, context);
    this.output(entry);
  }
  
  /**
   * Log error message
   */
  error(message: string, error?: Error, context?: Record<string, any>) {
    const entry = this.buildLogEntry(LogLevel.ERROR, message, context, error);
    this.output(entry);
  }
  
  /**
   * Log critical message
   */
  critical(message: string, error?: Error, context?: Record<string, any>) {
    const entry = this.buildLogEntry(LogLevel.CRITICAL, message, context, error);
    this.output(entry);
  }
  
  /**
   * Log API request
   */
  apiRequest(method: string, url: string, data?: any) {
    this.info('API Request', {
      type: 'api_request',
      method,
      url,
      data
    });
  }
  
  /**
   * Log API response
   */
  apiResponse(method: string, url: string, status: number, duration: number) {
    const level = status >= 400 ? LogLevel.ERROR : LogLevel.INFO;
    const entry = this.buildLogEntry(level, 'API Response', {
      type: 'api_response',
      method,
      url,
      status,
      duration_ms: duration
    });
    this.output(entry);
  }
  
  /**
   * Log API error
   */
  apiError(method: string, url: string, error: Error) {
    this.error('API Error', error, {
      type: 'api_error',
      method,
      url
    });
  }
  
  /**
   * Log user action for audit trail
   */
  audit(action: string, resource: string, resourceId?: number, data?: any) {
    this.info('User Action', {
      type: 'audit',
      action,
      resource,
      resourceId,
      data
    });
  }
  
  /**
   * Log performance metric
   */
  performance(metric: string, value: number, unit: string = 'ms') {
    this.info('Performance Metric', {
      type: 'performance',
      metric,
      value,
      unit
    });
  }
  
  /**
   * Set minimum log level
   */
  setLevel(level: LogLevel) {
    this.minLevel = level;
  }
}

// Global logger instance
export const logger = new Logger('store-frontend', LogLevel.INFO);

/**
 * Create a named logger instance
 */
export function getLogger(name: string, minLevel?: LogLevel): Logger {
  return new Logger(name, minLevel);
}

/**
 * Initialize logging with configuration
 */
export function initLogging(config?: { level?: LogLevel }) {
  if (config?.level) {
    logger.setLevel(config.level);
  }
  
  // Log application startup
  logger.info('Application Started', {
    environment: import.meta.env.MODE,
    version: import.meta.env.VITE_APP_VERSION || '1.0.0'
  });
  
  // Add global error handler
  window.addEventListener('error', (event) => {
    logger.critical('Uncaught Error', event.error, {
      filename: event.filename,
      lineno: event.lineno,
      colno: event.colno
    });
  });
  
  // Add unhandled promise rejection handler
  window.addEventListener('unhandledrejection', (event) => {
    logger.critical('Unhandled Promise Rejection', undefined, {
      reason: event.reason
    });
  });
  
  return logger;
}

/**
 * Axios interceptor for automatic API logging
 */
export function setupAxiosLogging(axiosInstance: any) {
  // Request interceptor
  axiosInstance.interceptors.request.use(
    (config: any) => {
      config.metadata = { startTime: Date.now() };
      logger.apiRequest(config.method.toUpperCase(), config.url, config.data);
      return config;
    },
    (error: Error) => {
      logger.apiError('REQUEST', 'unknown', error);
      return Promise.reject(error);
    }
  );
  
  // Response interceptor
  axiosInstance.interceptors.response.use(
    (response: any) => {
      const duration = Date.now() - response.config.metadata.startTime;
      logger.apiResponse(
        response.config.method.toUpperCase(),
        response.config.url,
        response.status,
        duration
      );
      return response;
    },
    (error: any) => {
      if (error.response) {
        const duration = Date.now() - error.config.metadata.startTime;
        logger.apiResponse(
          error.config.method.toUpperCase(),
          error.config.url,
          error.response.status,
          duration
        );
      } else {
        logger.apiError(
          error.config?.method?.toUpperCase() || 'UNKNOWN',
          error.config?.url || 'unknown',
          error
        );
      }
      return Promise.reject(error);
    }
  );
}

/**
 * React Error Boundary logger
 */
export function logErrorBoundary(error: Error, errorInfo: any) {
  logger.critical('React Error Boundary', error, {
    componentStack: errorInfo.componentStack
  });
}

export default logger;

File: backend/unneeded/src/utils/logger.py
# -*- coding: utf-8 -*-
"""
Comprehensive Logging System for Store Backend
================================================

Features:
- Structured JSON logging
- Multiple log levels with color coding
- Request/Response logging middleware
- Database query logging
- Performance metrics
- Error tracking with stack traces
- User action audit trail
- No file output - console/stdout only
"""

import logging
import sys
import json
import time
from datetime import datetime
from functools import wraps
from flask import request, g, has_request_context
from typing import Any, Dict, Optional
import traceback


class ColoredFormatter(logging.Formatter):
    """Custom formatter with color codes for console output."""

    # ANSI color codes
    COLORS = {
        "DEBUG": "\033[36m",  # Cyan
        "INFO": "\033[32m",  # Green
        "WARNING": "\033[33m",  # Yellow
        "ERROR": "\033[31m",  # Red
        "CRITICAL": "\033[35m",  # Magenta
        "RESET": "\033[0m",  # Reset
    }

    def format(self, record):
        """Format log record with colors."""
        color = self.COLORS.get(record.levelname, self.COLORS["RESET"])
        reset = self.COLORS["RESET"]

        # Add color to level name
        record.levelname = f"{color}{record.levelname}{reset}"

        return super().format(record)


class StructuredLogger:
    """
    Structured logger that outputs JSON-formatted logs to stdout.
    No file writing - designed for containerized environments.
    """

    def __init__(self, name: str = "store", level: str = "INFO"):
        self.logger = logging.getLogger(name)
        self.logger.setLevel(getattr(logging, level.upper()))

        # Remove existing handlers
        self.logger.handlers = []

        # Console handler with JSON formatting
        console_handler = logging.StreamHandler(sys.stdout)
        console_handler.setLevel(logging.DEBUG)

        # Use colored formatter for better readability
        formatter = ColoredFormatter(
            "%(asctime)s | %(levelname)s | %(name)s | %(message)s",
            datefmt="%Y-%m-%d %H:%M:%S",
        )
        console_handler.setFormatter(formatter)

        self.logger.addHandler(console_handler)

    def _build_log_entry(
        self,
        level: str,
        message: str,
        extra: Optional[Dict[str, Any]] = None,
        exc_info: Optional[Exception] = None,
    ) -> Dict[str, Any]:
        """Build structured log entry."""
        log_entry = {
            "timestamp": datetime.utcnow().isoformat(),
            "level": level,
            "message": message,
            "logger": self.logger.name,
        }

        # Add request context if available
        if has_request_context():
            log_entry["request"] = {
                "method": request.method,
                "path": request.path,
                "remote_addr": request.remote_addr,
                "user_agent": request.headers.get("User-Agent", "Unknown"),
            }

            # Add user info if available
            if hasattr(g, "current_user") and g.current_user:
                log_entry["user"] = {
                    "id": getattr(g.current_user, "id", None),
                    "username": getattr(g.current_user, "username", None),
                }

        # Add extra data
        if extra:
            log_entry["extra"] = extra

        # Add exception info
        if exc_info:
            log_entry["exception"] = {
                "type": type(exc_info).__name__,
                "message": str(exc_info),
                "traceback": traceback.format_exc(),
            }

        return log_entry

    def debug(self, message: str, extra: Optional[Dict] = None):
        """Log debug message."""
        log_entry = self._build_log_entry("DEBUG", message, extra)
        self.logger.debug(json.dumps(log_entry, ensure_ascii=False))

    def info(self, message: str, extra: Optional[Dict] = None):
        """Log info message."""
        log_entry = self._build_log_entry("INFO", message, extra)
        self.logger.info(json.dumps(log_entry, ensure_ascii=False))

    def warning(self, message: str, extra: Optional[Dict] = None):
        """Log warning message."""
        log_entry = self._build_log_entry("WARNING", message, extra)
        self.logger.warning(json.dumps(log_entry, ensure_ascii=False))

    def error(
        self,
        message: str,
        extra: Optional[Dict] = None,
        exc_info: Optional[Exception] = None,
    ):
        """Log error message with optional exception."""
        log_entry = self._build_log_entry("ERROR", message, extra, exc_info)
        self.logger.error(json.dumps(log_entry, ensure_ascii=False))

    def critical(
        self,
        message: str,
        extra: Optional[Dict] = None,
        exc_info: Optional[Exception] = None,
    ):
        """Log critical message with optional exception."""
        log_entry = self._build_log_entry("CRITICAL", message, extra, exc_info)
        self.logger.critical(json.dumps(log_entry, ensure_ascii=False))

    def audit(
        self,
        action: str,
        resource: str,
        resource_id: Optional[int] = None,
        extra: Optional[Dict] = None,
    ):
        """Log user action for audit trail."""
        audit_data = {
            "action": action,
            "resource": resource,
            "resource_id": resource_id,
        }

        if extra:
            audit_data.update(extra)

        log_entry = self._build_log_entry("INFO", "User Action", audit_data)
        log_entry["type"] = "audit"
        self.logger.info(json.dumps(log_entry, ensure_ascii=False))


# Global logger instance
app_logger = StructuredLogger(name="store", level="INFO")


def log_request_response(func):
    """Decorator to log request/response details."""

    @wraps(func)
    def wrapper(*args, **kwargs):
        # Log request
        start_time = time.time()

        app_logger.info(
            "Incoming Request",
            extra={
                "method": request.method,
                "path": request.path,
                "query_params": dict(request.args),
                "content_type": request.content_type,
            },
        )

        try:
            # Execute route handler
            response = func(*args, **kwargs)

            # Calculate duration
            duration = (time.time() - start_time) * 1000  # ms

            # Log response
            status_code = response[1] if isinstance(response, tuple) else 200
            app_logger.info(
                "Outgoing Response",
                extra={
                    "method": request.method,
                    "path": request.path,
                    "status_code": status_code,
                    "duration_ms": round(duration, 2),
                },
            )

            return response

        except Exception as e:
            duration = (time.time() - start_time) * 1000
            app_logger.error(
                "Request Failed",
                extra={
                    "method": request.method,
                    "path": request.path,
                    "duration_ms": round(duration, 2),
                },
                exc_info=e,
            )
            raise

    return wrapper


def log_db_query(query_type: str, table: str, duration_ms: float, **kwargs):
    """Log database query for monitoring."""
    app_logger.debug(
        "Database Query",
        extra={
            "query_type": query_type,
            "table": table,
            "duration_ms": round(duration_ms, 2),
            **kwargs,
        },
    )


def log_performance_metric(
    metric_name: str, value: float, unit: str = "ms", extra: Optional[Dict] = None
):
    """Log performance metric."""
    metric_data = {"metric": metric_name, "value": value, "unit": unit}

    if extra:
        metric_data.update(extra)

    log_entry = app_logger._build_log_entry("INFO", "Performance Metric", metric_data)
    log_entry["type"] = "performance"
    app_logger.logger.info(json.dumps(log_entry, ensure_ascii=False))


class RequestLoggingMiddleware:
    """
    Flask middleware to automatically log all requests and responses.
    """

    def __init__(self, app=None):
        if app:
            self.init_app(app)

    def init_app(self, app):
        """Initialize middleware with Flask app."""
        app.before_request(self.before_request)
        app.after_request(self.after_request)
        app.teardown_request(self.teardown_request)

    def before_request(self):
        """Log before request processing."""
        g.request_start_time = time.time()

        app_logger.debug(
            "Request Started",
            extra={
                "method": request.method,
                "path": request.path,
                "endpoint": request.endpoint,
            },
        )

    def after_request(self, response):
        """Log after request processing."""
        if hasattr(g, "request_start_time"):
            duration = (time.time() - g.request_start_time) * 1000

            app_logger.info(
                "Request Completed",
                extra={
                    "method": request.method,
                    "path": request.path,
                    "status_code": response.status_code,
                    "duration_ms": round(duration, 2),
                },
            )

        return response

    def teardown_request(self, exception=None):
        """Log request teardown and any errors."""
        if exception:
            app_logger.error(
                "Request Exception",
                extra={"method": request.method, "path": request.path},
                exc_info=exception,
            )


def init_logging(app, level: str = "INFO"):
    """
    Initialize logging for Flask application.

    Usage:
        from src.utils.logger import init_logging

        app = Flask(__name__)
        init_logging(app, level='INFO')
    """
    global app_logger

    # Set log level from config or parameter
    log_level = app.config.get("LOG_LEVEL", level).upper()
    app_logger = StructuredLogger(name="store", level=log_level)

    # Add request logging middleware
    RequestLoggingMiddleware(app)

    # Log startup
    app_logger.info(
        "Application Started",
        extra={
            "environment": app.config.get("ENV", "production"),
            "debug": app.config.get("DEBUG", False),
        },
    )

    return app_logger


# Export convenience functions
def get_logger(name: Optional[str] = None) -> StructuredLogger:
    """Get logger instance."""
    if name:
        return StructuredLogger(name=name)
    return app_logger


__all__ = [
    "app_logger",
    "get_logger",
    "init_logging",
    "log_request_response",
    "log_db_query",
    "log_performance_metric",
    "RequestLoggingMiddleware",
    "StructuredLogger",
]
